### ISSUE TITLE
Performance regression on Master daily build 20180608-09-15

### ISSUE THREAD

**Issue by jnativio** [istio/istio#6137]
Istio build: master-20180608-09-15
k8s version:
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.7-2+231cc32d0a1119", GitCommit:"231cc32d0a111940d45936b3206548c610869df2", GitTreeState:"clean", BuildDate:"2018-04-25T08:09:04Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

Regression Patrol detected a major performance regression on this build with the Full Istio configuration. The max throughput obtained was about 50% less than the previous build with the same configuration.
Full results can be seen here: 
https://ibmcloud-perf.istio.io/regpatrol/master-20180608-09-15/

The reason for this degradation was the excessive activity in the policy pod compared to previous builds. 
The charts below show the cpu usage by policy pod /containers during the run for the 0608 build:
![image](https://user-images.githubusercontent.com/18554027/41212195-26b416ee-6d0a-11e8-8a90-8d1aeea436cf.png)

And these charts are for the same metrics for the previous build (0531):
![image](https://user-images.githubusercontent.com/18554027/41212229-67293290-6d0a-11e8-97ba-1a2609ad4ef2.png)

**Comment by douglas-reid** [comment]
It looks like the previous build was 5-31.  So this reflects approx. one week's worth of changes.

**Comment by jnativio** [comment]
@douglas-reid That is correct, the previous build was from 05/31. Regpatrol watches builds published in https://gcsweb.istio.io/gcs/istio-prerelease/daily-build/ and these are the latest two.

**Comment by guptasu** [comment]
@jnativio for build number 5/31 the snapshot is taken on 5/31 or before that ?  I am trying to identify which PR potentially can cause the issue.
Secondly, any instructions on how to reproduce the issue

**Comment by geeknoid** [comment]
And if there's no easy way to repro this setup outside of the test environment, it is possible to generate and run intermediate builds so that we can do a binary search on commits to isolate the culprit?

**Comment by jnativio** [comment]
@guptasu I am not sure if I understood your question. The second snapshot was for the 05/31 build and it is an example. All other builds tested before showed a very low cpu footprint for the istio-policy pod.  As for reproducing the issue, it should not be difficult if you have a load generator. As last resort, I can run manually intermediate builds to help narrow down the time when the change happened.

**Comment by guptasu** [comment]
Thanks @jnativio for your offer and I think I will need your help in narrowing down the cause. I don't have the load generator. @jnativio I can go through the changes that went in from May 31st -> June 8th and create a set of intermediate mixer builders. It will be helpful if you can run the tests on the intermediate builds for us to narrow down the culprit PR. I will try my best to have the intermediate build count less. Thanks for your help.

**Comment by douglas-reid** [comment]
Those are mixer builds. Do we also need istio-proxy builds?  The charts seem to indicate heavy istio-proxy cpu usage as well.

**Comment by guptasu** [comment]
that is a good idea. I see that images for proxy is also available for the same SHAs
gcr.io/istio-testing/proxy:db97498d0c6c08812bd6ab3a98dfc4a0ae559933
gcr.io/istio-testing/proxy:dcaf9d7e4414c777c090665d08e0e6732b129bbf
gcr.io/istio-testing/proxy:a669ebfffcbd99e6ba0621cf2180264dcf709a5b

**Comment by douglas-reid** [comment]
Word on the street is that Mixer cache appears to be not working. This would result in a 10x increase in requests to `istio-policy`, which would explain CPU increase.

cc: @bianpengyuan

**Comment by mandarjog** [comment]
It turns out that Istio-policy mixer is actually doing more work.
Check cache is broken. https://ibmcloud-perf-grafana.istio.io/d/2/mixer-dashboard?from=1528573647000&to=1528576438000&orgId=6
Incoming request rate at Mixer is 1500 qps in the latest tests. It is normally less that 100 qps.

So Istio-policy is doing 15 times more work and has the cpu utilization to show it.

@bianpengyuan is debugging why cache hit rate is so low.

**Comment by ldemailly** [comment]
the issue is/should be reproducible using the simple [fortio setup](https://github.com/istio/istio/tree/master/tools#istio-load-testing-user-guide), see
https://fortio-daily.istio.io/?s=max.*s1_to_s2.*master

and we need to have more people using it to run their own tests rather than ask reg patrol to bissect

**Comment by jnativio** [comment]
Ok, so my understanding is that the root cause was found and we don't need to narrow down when the change was introduced.

**Comment by geeknoid** [comment]
Correct, no need for further investigation. It'll be a trivial fix and I'll
add a specific test case for this to prevent future regressions.
